diff --git a/ports/raspberrypi/supervisor/port.c b/ports/raspberrypi/supervisor/port.c
index 3b7b6d6f6d..3eb8df4e85 100644
--- a/ports/raspberrypi/supervisor/port.c
+++ b/ports/raspberrypi/supervisor/port.c
@@ -43,7 +43,6 @@
 #include "src/rp2_common/hardware_sync/include/hardware/sync.h"
 #include "src/rp2_common/hardware_timer/include/hardware/timer.h"
 #if CIRCUITPY_CYW43
-#include "py/mphal.h"
 #include "pico/cyw43_arch.h"
 #endif
 #include "src/common/pico_time/include/pico/time.h"
@@ -347,11 +346,6 @@ safe_mode_t port_init(void) {
     // Check brownout.
 
     #if CIRCUITPY_CYW43
-    // A small number of samples of pico w need an additional delay before
-    // initializing the cyw43 chip. Delays inside cyw43_arch_init_with_country
-    // are intended to meet the power on timing requirements, but apparently
-    // are inadequate. We'll back off this long delay based on future testing.
-    mp_hal_delay_ms(1000);
     // Change this as a placeholder as to how to init with country code.
     // Default country code is CYW43_COUNTRY_WORLDWIDE)
     if (cyw43_arch_init_with_country(PICO_CYW43_ARCH_DEFAULT_COUNTRY_CODE)) {
